### PR TITLE
Feature/workflow files

### PR DESCRIPTION
**Pull request by tiller1010**
https://werkbotstudios.teamwork.com/#/tasks/29567255

### Summary
- Added workflow flows from recent repos
- Added workflow files that behave like git flow finish commands
   - **publish_release_draft.yml**: publishes the draft release when "release/\*" is merged into "main", also publishes patches when a "hotfix/\*" branch is merged into a "support/\*" branch
   - **sync_develop_with_main.yml**: merges "main" into "develop" when main receives commits (merged PRs), also merges "support/\*" into "main" and main into "develop" when "support/\*" receives commits (merged PRs)

### Testing Steps
- [x] start a fork
- [x] confirm actions are enabled on your fork
- [x] merge this pr into your fork's "develop" branch
- [x] see my comment below and update your workflow file on "develop"
#### New release testing
- [x] from your develop branch, start a release: `git flow release start 1.1.0` **or** `git checkout -b release/1.1.0`
- [x] commit to your release branch if you like (release notes, comments, etc)
- [x] start a PR for your release branch into "main", use the appropriate semvar label ("minor" in the case of 1.1.0)
- [x] confirm this drafted a release
- [x] merge the PR with a merge commit, **DO NOT** delete the branch just yet. View the "Actions" tab, there should be a "publish" action running. You can delete your branch after this is finished
- [x] confirm the release was published
- [x] confirm "main" was automatically merged into "develop"
#### Support patch testing
- [x] start a support branch: `git flow support start 1.0.x 1.0.3` **or** `git checkout 1.0.3;git checkout -b support/1.0.x` (your last "1.0.x" tag may vary)
- [x] push this support branch up: `git push -u origin support/1.0.x`
- [x] start a hotfix branch off of your support branch: `git flow hotfix start 1.0.4 support/1.0.x` **or** `git checkout support/1.0.x;git checkout -b hotfix/1.0.4`
- [x] commit to your hotfix branch if you like (release notes, comments, etc)
- [x] start a PR for your hotfix branch into your support branch, use the appropriate semvar label ("patch" in the case of 1.0.4)
- [x] confirm this drafted a release
- [x] merge the PR with a merge commit, **DO NOT** delete the branch just yet. View the "Actions" tab, there should be a "publish" action running. You can delete your branch after this is finished
- [x] confirm the release was published (scroll down, as it is sorted lower)
- [x] confirm "support/1.0.x" was automatically merged into "main" and "main" was automatically merged into "develop"

### Issues/Concerns
- Git flow commands do not work so well if you have existing "hotfix/" or "release/" branches locally. These are not deleted automatically when things happen on github. I had to run `git branch -D ...` or `git flow release delete -f ...` a few times for the flow commands to work. Luckily, using these is not required, and flow branches can still be created normally.
- The publish action requires the PR branch to remain up until the action finishes. We may need to add a warning in the PR description for this (or any other git flow tips that could help now that I think about it).
- My test PRs for reference. It might be helpful to hover over these to see which branches went where:
https://github.com/tiller1010/silverstripe-spam-module/pulls?q=is%3Apr+is%3Aclosed

### CODE REVIEW

File: .github/workflows/publish_release_draft.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+jobs:
+  publish_release:
+    if: ${{ github.event.pull_request.merged }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Get release id
+        id: get_release_id
+        run: |
+          TOKEN=${{ secrets.GITHUB_TOKEN }}
+          RELEASEID=$(curl -H "Accept: application/vnd.github+json" -H "Authorization: token $TOKEN" https://api.github.com/repos/werkbot/silverstripe-spam-module/releases)
```
Comment (tiller1010) [inline .github/workflows/publish_release_draft.yml:20]: This is a raw curl call to the github actions api. You will need to update the `https://api.github.com/repos/werkbot/silverstripe-spam-module/releases` to your fork.

**Review by aletail — CHANGES_REQUESTED** [review]
This is pretty awesome, I put some thoughts below we can discuss further but interested in your thoughts:
- I can see a lot of stale branches coming from this process simply from forgetting to delete them
- After merging a PR, I could see the branch being deleted before the publish action is run - can we disable the branch delete? I will check settings in github as that could very well be a thing
- I could see the branch not being deleted after the publish action is run - can we automate deleting the branch after the publish action?
- The label checker should probably prevent any merging of a PR if it fails, I know I will forget this! (Also I have admin rights so I can merge whenever, this could be a github setting as well)
- It seems like "Merged main into develop" will come through on every release - which is not ideal, going to dirty up the commit history. Is there a way around this?

Again awesome stuff! I think if we can work through some of the above, to try to make if fool proof - it will be an awesome addition to our process!

**Comment by tiller1010** [comment]
@aletail 
> After merging a PR, I could see the branch being deleted before the publish action is run - can we disable the branch delete? I will check settings in github as that could very well be a thing

[Branch protection rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/managing-a-branch-protection-rule) prevent the delete button from showing, but it also prevents an action from deleting the branch later.

> It seems like "Merged main into develop" will come through on every release - which is not ideal, going to dirty up the commit history. Is there a way around this?

I will look into this, but a merge commit (default PR merge) created the least amount of repeated commits on subsequent PRs. When I used "Squash and commit" I would see several previous commits show up under later PRs.

**Comment by tiller1010** [comment]
@aletail 
> It seems like "Merged main into develop" will come through on every release - which is not ideal, going to dirty up the commit history. Is there a way around this?

In this example, I started a 1.3.x patch release, then I started a new 1.4 feature on develop, before finishing the 1.3.x release. When I finished the 1.3.x patch, the commit history still showed the real chronological development history: I started a new feature before finishing the release. I compared this to a local workflow in which I only use `git flow` commands, and the history was about the same.
gitflow-comparison

**Comment by tiller1010** [comment]
@aletail Here are some things I noted. I think we need to merge without creating a merge commit.

- A lot of "Merged main into develop" are showing on a [release PR into main](https://github.com/tiller1010/silverstripe-spam-module/pull/43).
	- These all come from previous hotfixes or releases since a merge commit is created when either is merged into "develop"
	- There is always at least one on "releases" into "main", since the release originated from "develop"
	- There are none with "hotfix" into "main", as these originate from "main" and not "develop"

- Previous release PR shows up [under the next release](https://github.com/tiller1010/silverstripe-spam-module/releases/tag/1.4.4)
	- This might be an issue caused by the merge commits mentioned above

**Review by aletail — APPROVED** [review]
This all worked really well, I can see a lot of conflicts popping up with the automatic merging from `main` -> `develop` etc...The manual process isn't too bad though, we will just need to figure a way these are properly being merged when those cases arise.

**Pull request merged**